File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.cpp
class Solution {
public:
  int lengthOfLongestSubstring(string s) {
    map<char, int> dp;
    int last = 0;
    int result = 0;
    for (string::size_type i = 0; i < s.size(); i++) {
      char ch = s[i];
      if (dp.count(ch) && dp[ch] >= last) {
        last = dp[ch] + 1;
      } else {
        if (i + 1 - last > result) {
          result = i + 1 - last;
        }
      }
      dp[ch] = i;
    }
    return result;
  }
};

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.java
class Solution {
  public int lengthOfLongestSubstring(String s) {
    HashMap<Character, Integer> dp = new HashMap<Character, Integer>(); 
    int last = 0;    
    int result = 0;
    for (int i = 0; i < s.length(); i++) {
      char ch = s.charAt(i);
      if (dp.containsKey(ch) && dp.get(ch) >= last) {
        last = dp.get(ch) + 1;
      } else {
        if (i - last + 1 > result) {
          result = i - last + 1;
        }
      }
      dp.put(ch, i);
    }
    return result;
  }
}